lib/auth/join_gitlab.go

```diff
@@ -61,24 +63,67 @@ func (a *Server) checkGitLabJoinRequest(ctx context.Context, req *types.Register
 	return claims, trace.Wrap(checkGitLabAllowRules(token, claims))
 }
 
+// joinRuleGlobMatch is used when comparing some rule fields from a
+// ProvisionToken  against a claim from a token. It allows simple pattern
+// matching:
+// - '*' matches zero or more characters.
+// - '?' matches any single character.
+// It returns true if a match is detected.
+func joinRuleGlobMatch(want string, got string) (bool, error) {
+	if want == "" {
+		return true, nil
+	}
+	pattern := regexp.QuoteMeta(want)
+	pattern = strings.ReplaceAll(pattern, `\*`, ".*")
+	pattern = strings.ReplaceAll(pattern, `\?`, ".")
+	pattern = "^" + pattern + "$"
+	matched, err := regexp.MatchString(pattern, got)
+	return matched, trace.Wrap(err)
```

Comment:
I think we have this sort of glob matching in some other place
Maybe here https://github.com/gravitational/teleport/blob/1e8cc61d872242c4a0abe2d1e68548d7a7d04215/lib/auth/join_iam.go#L238-L245

I think we should move this to a common place and be re-used in both places.

Reply:
317c695665